lib/openstack_query/query_parser.py

```diff
@@ -0,0 +1,211 @@
+from typing import List, Dict, Tuple, Union, Optional, Callable
+import logging
+from collections import OrderedDict
+from enums.query.props.prop_enum import PropEnum
+from openstack_query.handlers.prop_handler import PropHandler
+from custom_types.openstack_query.aliases import OpenstackResourceObj, PropValue
+from exceptions.query_property_mapping_error import QueryPropertyMappingError
+
+logger = logging.getLogger(__name__)
+
+
+class QueryParser:
+    """
+    Helper class for taking query output and parsing it into a format which can then be outputted.
+    Performs any sorting / grouping the user has specified
+    """
+
+    def __init__(self, prop_handler: PropHandler):
+        self._prop_handler = prop_handler
+        self._sort_by = {}
+        self._group_by = None
+        self._group_mappings = {}
+
+    @property
+    def group_by_prop(self):
+        return self._group_by
+
+    def _check_prop_valid(self, prop: PropEnum) -> None:
+        """
+        method which checks if the given property is valid - i.e. has an associated function mapping in
+        self.prop_handler which takes a openstack resource and returns the corresponding property for that object
+        :param prop: An enum representing the desired property
+        """
+        if not self._prop_handler.check_supported(prop):
+            raise QueryPropertyMappingError(
+                "Error: failed to get property mapping, property is not supported by prop_handler"
+            )
+
+    def parse_sort_by(self, *sort_by: Tuple[PropEnum, bool]) -> None:
+        """
+        Public method used to configure sorting results
+        :param sort_by: one or more tuples of property name to sort by and order
+            - set order to True for descending, False for ascending
+        """
+        self._sort_by = {}
+        for prop, order in sort_by:
+            self._check_prop_valid(prop)
+            order_log_str = "DESC" if order else "ASC"
+            logging.debug(
+                "adding sorting params: %s, order: %s", prop.name, order_log_str
+            )
+            self._sort_by.update({prop: order})
+
+        logging.debug(
+            "sort params: %s",
+            "\n\t".join(
+                f"{prop}, order: {'DESC' if order else 'ASC'}"
+                for prop, order in self._sort_by.items()
+            ),
+        )
+
+    def parse_group_by(
```

Comment:
There's enough grouping stuff I'd argue that it's worth just using the `refactor -> extract class` to extract out `OutputGrouper` or something similar from `query_parser`

Reply:
I think it is fine for now - but I made a separate issue